### ISSUE TITLE
Use raw input instead of hooking to keyboard and mouse (Windows)

### ISSUE THREAD

**Issue by davidebeatrici** [mumble-voip/mumble#4039]
Several users complained about the global shortcuts engine, mainly because it affects the keyboard and mouse behavior for all applications.

More specifically, it reportedly causes the polling rate to drop significantly, which is absolutely unacceptable (#3275).

Also, using the debugger's "pause" function is nearly impossible: Windows has to wait for the callback to timeout before registering the input, causing a consistent and heavy slowdown (#3553).

A possible solution consists in using raw input: https://docs.microsoft.com/en-us/windows/win32/inputdev/raw-input

Another advantage of raw input is that we can treat multiple keyboards and mice as separate devices.

**Comment by trudnorx** [comment]
Why can't just use RegisterHotKey or other stuff for this? Why would you have to be hooking the mouse/keyboard or using raw input? Why would you have to be directly accessing the mouse/keyboard and be causing potential for any problem?

**Comment by davidebeatrici** [comment]
[RegisterHotKey()](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-registerhotkey) can definitely be used, but it's not ideal for multiple reasons:

- Performance. As explained in Microsoft's documentation, Windows searches for registered hotkeys each time a key is pressed.
The operation probably only takes a few milliseconds, but it's something that can be entirely avoided.
- No support for HID devices other than mouse and keyboard.
- We cannot associate a hotkey with a window created by another thread, thus we would have to do it in `MainWindow` rather than `GlobalShortcutWin`.

Hooking to the keyboard and mouse has the potential to cause issues because we're between the devices and the rest of the system.

By using raw input, we just receive a `WM_INPUT` message that is then processed in one of our threads.

**Comment by davidebeatrici** [comment]
For reference: https://blog.molecular-matters.com/2011/09/05/properly-handling-keyboard-input

**Comment by SizzlingCalamari** [comment]
A temporary solution could be to revert https://github.com/mumble-voip/mumble/pull/1559.
Mouse and kb callbacks are run in the context of the thread that hooked. This means callbacks are run in a normal priority process with a normal priority thread, which can lag the system if not scheduled quickly enough.

**Comment by SizzlingCalamari** [comment]
Actually, code says we're setting the GlobalShortcutWin thread to time critical: https://github.com/mumble-voip/mumble/blob/82cfcb84a59b9b66dba0dbdf6d877fb8e2451813/src/mumble/GlobalShortcut_win.cpp#L191

Win32 docs says the priority of a time critical thread from a normal vs high priority process are the same (15). https://docs.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities
Perhaps not the issue then (needs investigation)?

Regardless, thought of this issue came from this recent post https://www.teamfortress.tv/58146/moving-mouse-causes-tf2-to-drop-20-30-fps. Issue seems to be fixed by using v1.2.19 and the issue occurs when using 1.3.0-rc1 (which is when my priority change occurred).

I'm trying to contact the OP to test the two versions of mumble alone, without any of the suggestions that others have posted.

**Comment by asantoni** [comment]
Even though you're setting the thread priority to TimeCriticalPriority, it could still be affected by priority inversion (the thread waiting on a lower priority thread). Either way, as a design principal, Mumble should not be blocking global processing of the mouse and keyboard because it interferes with other applications (like this stuttering bug). Handling of keyboard and mouse events in Mumble is not time critical - it should be handled asynchronously in a separate thread, regardless of whether you use raw input or the current approach.

**Comment by hauk92** [comment]
I have got a Windows installation where the mouse cursor starts lagging after I open Mumble. This issue is still present with version 1.2.19 of Mumble.

**Comment by Krzmbrzl** [comment]
> This issue is still present with version 1.2.19 of Mumble.

Well that version (or rather the 1.2.x series as a whole) is pretty much ancient. You should probably consider upgrading to 1.3.x (1.3.3 is currently the latest release).
The support for raw input is not included in 1.3.x though. It will probably end up in the 1.4.x series but as for now it still needs to be implemented (though @davidebeatrici is working on it).

**Comment by hauk92** [comment]
I am not actively using that version. I installed it in response to the workaround and Team Fortress comment, sorry I should have mentioned that. That said my issue is present in version 1.2.3 (I did not go back further), 1.2.19 and 1.3.3.

**Comment by SizzlingCalamari** [comment]
That aligns with what I found from the OP of that thread and a few others. Older versions did not fix the issue. OP reduced mouse usb polling rate, which would have made the issue way less noticeable.

The only workaround solution I've found to this issue is to disable cpu c-states in the bios. C-states are low power states per cpu core. There's measurable overhead when cpu cores switch between these low power states. Interrupts and periodic code execution requiring low latency are particularly affected when this power saving feature is enabled. 

Intel does say in its cpu manuals that c-states should be enabled to ensure cpu longevity, but I think it's worth the tradeoff for higher system performance.

@hauk92 are you able to try disabling C-States in the bios?

**Comment by hauk92** [comment]
Yes, I now tried disabling C-states, fixing the CPU frequency to 3 GHz, disabling hyper-threading, setting the polling rate of the mouse to 125Hz in Logitech G HUB, deinstalling Logitech G HUB and using different mice but the issue still persists.

**Comment by GoodFightt** [comment]
anyone find a solution? without mumble my polling rate is around 650 HZ. With mumble, it drops to 400HZ. I am assuming this is why my mouse glitches around in a game while I use mumble. How do I fix this?

**Comment by davidebeatrici** [comment]
https://dl.mumble.info/reg/disable-winhooks.reg

**Comment by hauk92** [comment]
Nice, this works for me, there is even a setting under shortcuts to disable Windows hooks. Thank you!

I had a look at a few similar issues and a few people reported that switching USB ports for certain devices works for them.
Plugging my keyboard from the USB 2 ports of my motherboard into the USB 3 ports also fixes the issue for me.

Hardware:
MSI B450-A PRO MAX
Ryzen 2700
Das Keyboard Model S

Another system with the same keyboard but a different MSI B450 board and a Ryzen 3000 CPU does not have these problems.
This is probably partially the fault of my fairly old Windows installation, which already migrated from a SandyBridge motherboard...

**Comment by GoodFightt** [comment]
[image: image.png]
Where?

**Comment by GoodFightt** [comment]
![image](https://user-images.githubusercontent.com/61664595/105415074-5032d580-5c06-11eb-843b-c9954629e0e1.png)
Disabling this will fix everything, thank me later! <3
